### PR TITLE
Fixes #3640 by shuffling endpoints in the round-robin load balancer

### LINKED ISSUE #3640: Round robin load balancer can overwhelm first pod behind service
This is particularly easy to do with a new cluster.

Create a service `S` and deploy `N` pods behind it. Start up `M` pods which create a persistent connection to service `S`.

Because the round robin load balancer always starts at the 0th endpoint, and the endpoint list is consistent cluster-wide, you will see that you now have `M` connections to Pod 1, and 0 connections to all other pods 2 through N.

Perhaps the the round robin load balancer should start off at a random index [0,N) when the list of endpoints changes.

(Concretely: we had 3 instances of twemproxy behind a "redis" service, fired up a bunch of worker pods (which happen to connect once, do all of their work, then exit), and each worker pod connected to the same instance of twemproxy, with the other 2 remaining completely idle.)

### PR DESCRIPTION
**Pull request by smreed**
Starts the round-robin index at 0, but shuffles the endpoints OnUpdate to prevent every minion in the cluster from pegging the first pod behind a service at the same time.

I'm not super excited about how I updated the unit tests, but they pass.

### CODE REVIEW

**Comment by bgrant0607** [comment]
Thanks! Would it be possible for you to run the e2e test? It requires a cloud provider that supports multiple minions (so, not local-up-cluster.sh).

https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md

Otherwise, LGTM.

/cc @thockin

**Comment by smreed** [comment]
Sure thing, running the e2e tests now.

**Comment by smreed** [comment]
One failure, the kubernetes RO test failed with a "rate limit exceeded" error. @bgrant0607 Should I re-run?

    I0120 17:05:02.981654   10521 driver.go:137] Running test 5 TestKubernetesROService
    E0120 17:05:03.034754   10521 roservice.go:33] unexpected error listing services using ro service: request [&{Method:GET URL:https://130.211.114.218/api/v1beta1/proxy/services/kubernetes-ro/api/v1beta1/services?namespace=default Proto:HTTP/1.1 ProtoMajor:1 ProtoMinor:1 Header:map[] Body:<nil> ContentLength:0 TransferEncoding:[] Close:false Host:130.211.114.218 Form:map[] PostForm:map[] MultipartForm:<nil> Trailer:map[] RemoteAddr: RequestURI: TLS:<nil>}] failed (503) 503 Service Unavailable: Rate limit exceeded.
    I0120 17:05:03.034818   10521 driver.go:140]         test 5 failed
    ...
    I0120 17:06:46.794423   10521 driver.go:43] 1..9
    I0120 17:06:46.794427   10521 driver.go:46] ok 1 - TestPodUpdate
    I0120 17:06:46.794431   10521 driver.go:46] ok 2 - TestPodHasServiceEnvVars
    I0120 17:06:46.794434   10521 driver.go:46] ok 3 - TestBasic
    I0120 17:06:46.794437   10521 driver.go:46] ok 4 - TestClusterDNS
    I0120 17:06:46.794441   10521 driver.go:48] not ok 5 - TestKubernetesROService
    I0120 17:06:46.794444   10521 driver.go:46] ok 6 - TestKubeletSendsEvent
    I0120 17:06:46.794447   10521 driver.go:46] ok 7 - TestPrivate
    I0120 17:06:46.794450   10521 driver.go:46] ok 8 - TestNetwork
    I0120 17:06:46.794453   10521 driver.go:46] ok 9 - TestImportantURLs
    F0120 17:06:46.794457   10521 driver.go:151] At least one test failed
    2015/01/20 17:06:46 Error running goe2e.sh: exit status 255
    not ok 7 - goe2e.sh
    ...
    2015/01/20 17:10:16 Passed tests: basic.sh[1/1] certs.sh[1/1] guestbook.sh[1/1] liveness.sh[1/1] monitoring.sh[1/1] pd.sh[1/1] private.sh[1/1] services.sh[1/1] update.sh[1/1]
    2015/01/20 17:10:16 Flaky tests:
    2015/01/20 17:10:16 Failed tests: goe2e.sh[0/1]
    2015/01/20 17:10:16 1 test(s) failed.

**Comment by bgrant0607** [comment]
For now, yes, please try again. I'll file some issues about this failure.

**Comment by smreed** [comment]
Another run through and the test that failed above passes, but others have failed. The 100% successful e2e test run eludes me. Do larger GCE instances help? I'll give that a shot.

**Comment by smreed** [comment]
2015/01/20 21:02:32 Passed tests: basic.sh[1/1] certs.sh[1/1] goe2e.sh[1/1] guestbook.sh[1/1] liveness.sh[1/1] monitoring.sh[1/1] pd.sh[1/1] update.sh[1/1]
    2015/01/20 21:02:32 Flaky tests:
    2015/01/20 21:02:32 Failed tests: private.sh[0/1] services.sh[0/1]
    2015/01/20 21:02:32 2 test(s) failed.

**Comment by bgrant0607** [comment]
Sadly e2e is very flaky these days, I believe. Maybe @thockin or @satnam6502 has suggestions.

The previous failure doesn't look like it could be due to this change.

What is the services.sh failure?

**Comment by smreed** [comment]
services.sh was, I believe, a timeout waiting for "endpoints to propagate"

**Comment by smreed** [comment]
```
Waiting for endpoints to propagate
Checking if 4ac53e79-a12e-11e4-b3bf-42010af0e0b34ac57a88-a12e-11e4-b3bf-42010af0e0b34ac69370-a12e-11e4-b3bf-42010af0e0b3  == 4ac53e79-a12e-11e4-b3bf-42010af0e0b3 4ac57a88-a12e-11e4-b3bf-42010af0e0b3 4ac69370-a12e-11e4-b3bf-42010af0e0b3
 
Waiting for endpoints to propagate
Checking if 4ac53e79-a12e-11e4-b3bf-42010af0e0b34ac57a88-a12e-11e4-b3bf-42010af0e0b34ac69370-a12e-11e4-b3bf-42010af0e0b3  == 4ac53e79-a12e-11e4-b3bf-42010af0e0b3 4ac57a88-a12e-11e4-b3bf-42010af0e0b3 4ac69370-a12e-11e4-b3bf-42010af0e0b3
 
Waiting for endpoints to propagate
Checking if 4ac53e79-a12e-11e4-b3bf-42010af0e0b34ac57a88-a12e-11e4-b3bf-42010af0e0b34ac69370-a12e-11e4-b3bf-42010af0e0b3  == 4ac53e79-a12e-11e4-b3bf-42010af0e0b3 4ac57a88-a12e-11e4-b3bf-42010af0e0b3 4ac69370-a12e-11e4-b3bf-42010af0e0b3
 
Waiting for endpoints to propagate
Endpoints did not propagate in time
```

**Comment by smreed** [comment]
Those strings seem to be the same minus some whitespace delimiters in one and not the other. :/ Bug in e2e?

**Comment by satnam6502** [comment]
There was a short period when private.sh would not have worked because of
changes to the image it tested -- but I think it should be fixed now.
Please let me know if you see any failures involving basic.sh, private.sh
or in the Go tests for TestBasic or TestPrivate.

Cheers,

Satnam

**Comment by bgrant0607** [comment]
Huh. Seems likely.

I'm running e2e on head. I'll let you know whether I see the same types of failures. Regardless, none of these failures seem like they could be caused by this change.

**Comment by smreed** [comment]
Latest run:

```
2015/01/20 21:40:07 Passed tests: basic.sh[1/1] certs.sh[1/1] goe2e.sh[1/1] guestbook.sh[1/1] liveness.sh[1/1] monitoring.sh[1/1] pd.sh[1/1] private.sh[1/1] update.sh[1/1]
2015/01/20 21:40:07 Flaky tests:
2015/01/20 21:40:07 Failed tests: services.sh[0/1]
2015/01/20 21:40:07 1 test(s) failed.
```

Same services.sh failure (waiting for endpoints to propagate, but it prints what look like equivalent lists of pod ids).

**Comment by smreed** [comment]
After getting the same endpoints failure in the service.sh e2e test I'm starting to think that my random ordering of the endpoints is affecting this test. I think the e2e test isn't properly sorting the pod IDs, so I'm going to spend some time looking at fixing that. I'll add to this PR if I find anything.

**Comment by erictune** [comment]
what cloud provider are you using?  services.sh doesn't work on vagrant,
FYI.

**Comment by smreed** [comment]
GCE

**Comment by smreed** [comment]
I think the container that serves the hostname changed and isn't writing a trailing newline, which is breaking my e2e tests.

**Comment by smreed** [comment]
Updated PR to include hack to services.sh that gets my e2e tests passing. I tested it out with a duplicate echo (in case the serve-hostname container output changes) and the tests still passed, so this should be safe. Happy to back that out thought if you'd rather see that in another PR.

TL;DR e2e tests pass 100%

@bgrant0607 PTAL

File: hack/e2e-suite/services.sh [hunk 1]
```diff
@@ -252,7 +253,8 @@ function verify_from_container() {
             ok=false
             for j in $(seq -s' ' 1 10); do
               if wget -q -T 1 -O - http://$2:$3; then
-		ok=true
```
Comment (smreed) [inline hack/e2e-suite/services.sh]: This line contained tabs, which is why it changes here.

File: pkg/proxy/roundrobin.go [hunk 2]
```diff
@@ -210,6 +221,10 @@ func (lb *LoadBalancerRR) OnUpdate(endpoints []api.Endpoints) {
 	for _, endpoint := range endpoints {
 		existingEndpoints, exists := lb.endpointsMap[endpoint.Name]
 		validEndpoints := filterValidEndpoints(endpoint.Endpoints)
+		// Need to compare sorted endpoints here, since they are shuffled below
+		// before being put into endpointsMap
+		sort.Strings(existingEndpoints)
```
Comment (thockin) [inline pkg/proxy/roundrobin.go]: I am pretty sure that, given the reference-nature of slices, this will actually sort the lb.endpointsMap[endpoint.Name] slice, thus undoing your shuffle.

If we keep this general approach, you need to sort-and-copy the slice.

Comment (smreed) [inline pkg/proxy/roundrobin.go]: Yes, you're right. This would undo the shuffle in cases where the endpoints don't change. I'll fix.

File: hack/e2e-suite/services.sh [hunk 3]
```diff
@@ -206,6 +206,7 @@ function wait_for_service_up() {
         set -e;
         for i in $(seq -s' ' 1 $4); do
           curl -s --connect-timeout 1 http://$2:$3;
+          echo
```
Comment (bgrant0607) [inline hack/e2e-suite/services.sh]: FWIW, the other lines end in semi-colons. Please make this consistent.

File: pkg/proxy/roundrobin.go (hunk 2 above)
Comment (bgrant0607) [inline pkg/proxy/roundrobin.go]: This should be easy to confirm. If we need to copy:
http://blog.golang.org/go-slices-usage-and-internals

Comment (thockin) [inline pkg/proxy/roundrobin.go]: I *know* we alias slices all over the place, and we're just begging for this bug to pop up other places.

File: hack/e2e-suite/services.sh (hunk 1 above)
Comment (bgrant0607) [inline hack/e2e-suite/services.sh]: Could you please fork these fixes into another PR? I'd like to get them in asap.

File: hack/e2e-suite/services.sh (hunk 3 above)
Comment (thockin) [inline hack/e2e-suite/services.sh]: Good catch - this is within an ssh argument string.

File: pkg/proxy/roundrobin.go (hunk 2 above)
Comment (smreed) [inline pkg/proxy/roundrobin.go]: Yeah a copy is definitely required to prevent `sort.Strings` from undoing the shuffle.

http://play.golang.org/p/OuFn_s6mJZ

File: hack/e2e-suite/services.sh (hunk 1 above)
Comment (smreed) [inline hack/e2e-suite/services.sh]: Will do (with semi-colons where you asked for them)

File: pkg/proxy/roundrobin.go (hunk 2 above)
Comment (thockin) [inline pkg/proxy/roundrobin.go]: Could we avoid all this by just choosing a random index of the slice, rather than literally round-robin?

Comment (smreed) [inline pkg/proxy/roundrobin.go]: It would change the behavior subtly in that the same pod could get two connections in a row. Not sure if that's a big deal or not.

File: pkg/proxy/roundrobin.go [hunk 4]
```diff
@@ -210,14 +236,14 @@ func (lb *LoadBalancerRR) OnUpdate(endpoints []api.Endpoints) {
 	for _, endpoint := range endpoints {
 		existingEndpoints, exists := lb.endpointsMap[endpoint.Name]
 		validEndpoints := filterValidEndpoints(endpoint.Endpoints)
-		if !exists || !reflect.DeepEqual(existingEndpoints, validEndpoints) {
+		if !exists || !endpointsAreEqual(existingEndpoints, validEndpoints) {
```
Comment (smreed) [inline pkg/proxy/roundrobin.go]: Changed how/where slices are sorted and compared. @thockin PTAL

Comment (thockin) [inline pkg/proxy/roundrobin.go]: Now you have pessimized - validEndpoints is wholly owned at this point.  Can we do something like:
```
// We own validEndpoints but existingEndpoints is a reference to the stored data.
if !exists || !reflect.DeepEqual(util.SortSlice(util.CopySlice(existingEndpoints)), util.SortSlice(validEndpoints)) {
```

I am dismayed by Go's sharp edges and lack of helpful functions here.

Comment (smreed) [inline pkg/proxy/roundrobin.go]: To be clear, are you suggesting I add SortSlice and CopySlice to github.com/GoogleCloudPlatform/kubernetes/pkg/util?

I get your point, that you want to copy the unnecessary copy of `validEndpoints`. I also looked a bit into `util.StringSet`. I could call `HasAll` and check the lengths.

Comment (thockin) [inline pkg/proxy/roundrobin.go]: That is what I was suggesting.  I thought about StringSet, too, but there's no way I can see to deterministically iterate it in order for a given shuffle.  It's a set, not a list.

I think something like `SortSlice([]string) []string` and `CopySlice([]string) []string` would be nice.  Maybe even cleaner to do pkg/util/slice and make it `slice.SortStrings([]string) []string)` and `slice.CopyStrings([]string) []string` (so we can add more later if needed).

Another attack would be to make something like `util.IndexStrings(slice []string) []*string` which would do the sort on pointers to strings, to avoid copying the input.

Comment (smreed) [inline pkg/proxy/roundrobin.go]: Sounds fun, I'll hopefully be able to put something together by tomorrow.

**Comment by bgrant0607** [comment]
FYI, TestKubernetesROService is failing at head. It's not this change.

**Comment by satnam6502** [comment]
Also, regarding the missing newline, apart from a short period when `kubernetes/serve_hostname` was accidentally made equal to `kubernetes/serve_hostname:1.0` I think you should observe the previous behaviour with `kubernetes/serve_hostname` (i.e. with a newline at the end of the hostname) whereas `kubernetes/serve_hostname:1.0` does not have the newline.

**Comment by bgrant0607** [comment]
And, there was a period where e2e was failing due to expecting more minions than were created. Sigh.

Our latest Jenkins run of e2e passed, and we disabled TestKubernetesROService. Update and try again.

**Comment by brendandburns** [comment]
Given that this is an outside contribution essential by a volunteer, I'm
willing to take a larger risk on merging than average.  Can we just merge
here and deal with the breaks if they occur?  I think Steve has already
gone above and beyond the call in running e2e here...

Alternatively I can take over running e2e for this PR...

Brendan

File: pkg/proxy/roundrobin.go (hunk 4 above)
Comment (thockin) [inline pkg/proxy/roundrobin.go]: Sounds good.  Poke this PR when you're ready.

**Comment by thockin** [comment]
I don't think it's quite ready to commit yet, but it's close.

File: pkg/proxy/roundrobin.go (hunk 4 above)
Comment (smreed) [inline pkg/proxy/roundrobin.go]: `slice.SortStrings([]string) []string` is a weird one in that it sorts in place, and also returns a reference to the slice. It makes it a somewhat strange "util" method and I would worry that it'll trip people up. If you think that's fine, I'll continue on.

Comment (thockin) [inline pkg/proxy/roundrobin.go]: Just document it.  The fact that it returns a reference to its input is what makes it useful in functional-style ways.  It's a well-established pattern in lots of languages - just document that it does an in-place sort.  Or maybe call it SortStringsInPlace() if it makes you more comfortable

**Comment by smreed** [comment]
Whoa, whoops. Did I rebase when I should've merged? 

Anyway, @thockin PTAL, this latest version uses slice util functions to do the copying and sorting, keeping the roundrobin code change to a minimum.

**Comment by smreed** [comment]
I'm going to close and re-open this, I didn't mean to push all of these commits into the branch. :/

**Comment by thockin** [comment]
LOL
